src/components/tl/ucp/tl_ucp_coll.h

```diff
@@ -121,6 +126,27 @@ typedef struct ucc_tl_ucp_task {
             ucc_ee_executor_task_t *etask;
             ucc_ee_executor_t      *executor;
         } allreduce_kn;
+        struct {
+            int                                        reduce_in_progress;
+            ucp_rkey_h                                *src_rkeys; //unpacked
+            ucp_rkey_h                                *dst_rkeys; //unpacked
+            ucp_ep_h                                  *eps;
```

Comment:
why need eps here?

Reply:
We're deliberately storing EPs to avoid going through the eps looking function, we can remove it, but I'm afraid to introduce yet another indirection where we can just have a rank like lookup.

I would prefer to keep it. Again we're not creating additional EPs, just storing for faster lookup.

Reply:
For faster/cleaner lookup. These eps arent duplicated, they just stored during coll start so that we dont need to call tl_ucp_get_ep in the progress function.

Reply:
ok, but IMHO for every collective allocating an array of eps, do memcopy from hash table to that array and deallocating it is more costly than query ep from hash. If I remember correctly for hash table of 1000 eps query time is 0.02us

Reply:
here we have only one memcpy per hash though. without it we will lookup the same hash multiple times per ep.